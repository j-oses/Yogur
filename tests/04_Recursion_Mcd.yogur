// Calculates GCD. Useful to check that recursion behaves correctly.

def mod(a: Int, b: Int) -> c: Int {
	c = a - (a / b) * b
}

def mcd(a: Int, b: Int) -> c: Int {
	if b == 0 {
		c = a
	} else {
		c = mcd(b, mod(a, b))
	}
}

var a: Int = mcd(7, 15)