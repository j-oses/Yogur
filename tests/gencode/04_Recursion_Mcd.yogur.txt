 {0}    ssp 1;
 {1}    ujp 18; \\ endFun1
\\ def mod
\\ fun1:
 {2}    ssp 7;
 {3}    sep 3;
 {4}    lda 0 0;
 {5}    lda 0 5;
 {6}    ind;
 {7}    lda 0 5;
 {8}    ind;
 {9}    lda 0 6;
 {10}   ind;
 {11}   div;
 {12}   lda 0 6;
 {13}   ind;
 {14}   mul;
 {15}   sub;
 {16}   sto;
 {17}   retf;
\\ endFun1:
 {18}   ujp 44; \\ endFun2
\\ def mcd
\\ fun2:
 {19}   ssp 7;
 {20}   sep 2;
 {21}   lda 0 6;
 {22}   ind;
 {23}   ldc 0;
 {24}   equ;
 {25}   fjp 31; \\ else3
 {26}   lda 0 0;
 {27}   lda 0 5;
 {28}   ind;
 {29}   sto;
 {30}   ujp 43; \\ endif4
\\ else3:
 {31}   lda 0 0;
 {32}   mst 1;
 {33}   lda 0 6;
 {34}   ind;
 {35}   mst 1;
 {36}   lda 0 5;
 {37}   ind;
 {38}   lda 0 6;
 {39}   ind;
 {40}   cup 2 2; \\ fun1
 {41}   cup 2 19; \\ fun2
 {42}   sto;
\\ endif4:
 {43}   retf;
\\ endFun2:
 {44}   lda 0 0;
 {45}   mst 0;
 {46}   ldc 7;
 {47}   ldc 15;
 {48}   cup 2 19; \\ fun2
 {49}   sto;
 {50}   stp;
