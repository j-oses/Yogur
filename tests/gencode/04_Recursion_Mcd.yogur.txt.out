Run successfully completed
Final P-machine state:
PC = 50   MP = 0   SP = 0   EP = -1   NP = 100   code size = 51
STORE = 
(0,Int 1)
(1,Int 0)
(2,Int 1)
(3,Int 0)
(4,Int 0)
(5,Int (-1))
(6,Int 49)
(7,Int 7)
(8,Int 15)
(9,Int 2)
(10,Int 1)
(11,Int 0)
(12,Int 2)
(13,Int 22)
(14,Int 42)
(15,Int 15)
(16,Int 7)
(17,Int 10)
(18,Int 1)
(19,Int 0)
(20,Int 10)
(21,Int 30)
(22,Int 42)
(23,Int 7)
(24,Int 1)
(25,Int 18)
(26,Int 1)
(27,Int 0)
(28,Int 18)
(29,Int 38)
(30,Int 42)
(31,Int 1)
(32,Int 0)
(33,Int 26)
(34,Int 1)
(35,Int 38)
(36,Int 41)
(37,Int 7)
(38,Int 1)
(39,Int 32)
(40,Int 0)
(41,Int 7)
(42,Int 1)
(43,Null)
(44,Null)
(45,Null)
(46,Null)
(47,Null)
(48,Null)
(49,Null)
(50,Null)
(51,Null)
(52,Null)
(53,Null)
(54,Null)
(55,Null)
(56,Null)
(57,Null)
(58,Null)
(59,Null)
(60,Null)
(61,Null)
(62,Null)
(63,Null)
(64,Null)
(65,Null)
(66,Null)
(67,Null)
(68,Null)
(69,Null)
(70,Null)
(71,Null)
(72,Null)
(73,Null)
(74,Null)
(75,Null)
(76,Null)
(77,Null)
(78,Null)
(79,Null)
(80,Null)
(81,Null)
(82,Null)
(83,Null)
(84,Null)
(85,Null)
(86,Null)
(87,Null)
(88,Null)
(89,Null)
(90,Null)
(91,Null)
(92,Null)
(93,Null)
(94,Null)
(95,Null)
(96,Null)
(97,Null)
(98,Null)
(99,Null)
